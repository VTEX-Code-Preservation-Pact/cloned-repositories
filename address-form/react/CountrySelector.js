import React, { Component } from 'react'
import PropTypes from 'prop-types'

import AddressShapeWithValidation from './propTypes/AddressShapeWithValidation'
import InputFieldContainer from './InputFieldContainer'
import DefaultInput from './inputs/DefaultInput'
import { injectAddressContext } from './addressContainerContext'
import { intlShape, injectIntl } from './intl/utils'

class CountrySelector extends Component {
  handleChangeCountry = (changedFields) => {
    const { address } = this.props

    if (changedFields.country.value === address.country.value) return

    this.props.onChangeAddress({
      country: { value: changedFields.country.value },
      city: { value: null },
      complement: { value: null },
      geoCoordinates: { value: [] },
      neighborhood: { value: null },
      number: { value: null },
      postalCode: { value: null },
      reference: { value: null },
      state: { value: null },
      street: { value: null },
    })
  }

  shouldComponentUpdate(prevProps) {
    if (prevProps.address.country.value !== this.props.address.country.value) {
      return true
    }

    if (prevProps.Input !== this.props.Input) {
      return true
    }

    if (prevProps.shipsTo !== this.props.shipsTo) {
      return true
    }

    return false
  }

  sortOptionsByLabel(options) {
    return options.slice().sort((a, b) => a.label.localeCompare(b.label))
  }

  render() {
    const { address, shipsTo, Input, intl } = this.props
    const field = {
      name: 'country',
      label: 'country',
      optionsCaption: false,
      value: address.country.value,
    }

    return (
      <InputFieldContainer
        intl={intl}
        Input={Input}
        field={field}
        rules={{}}
        options={this.sortOptionsByLabel(shipsTo)}
        address={address}
        onChangeAddress={this.handleChangeCountry}
      />
    )
  }
}

CountrySelector.defaultProps = {
  Input: DefaultInput,
}

CountrySelector.propTypes = {
  Input: PropTypes.func,
  intl: intlShape,
  address: AddressShapeWithValidation,
  shipsTo: PropTypes.array.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

export default injectAddressContext(injectIntl(CountrySelector))
