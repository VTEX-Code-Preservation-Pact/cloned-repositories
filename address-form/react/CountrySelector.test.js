import React from 'react'
import { render, mount, rendererCreate } from 'test-utils'

import CountrySelector from './CountrySelector'
import newAddress from './__mocks__/newAddress'

describe('CountrySelector', () => {
  const shipsTo = [
    { value: 'BRA', label: 'Brazil' },
    { value: 'USA', label: 'United States of America' },
  ]

  it('renders without crashing', () => {
    expect(() =>
      render(
        <CountrySelector
          address={newAddress}
          shipsTo={shipsTo}
          onChangeAddress={jest.fn()}
        />
      )
    ).not.toThrow()
  })

  it('show options', () => {
    const tree = rendererCreate(
      <CountrySelector
        address={newAddress}
        shipsTo={shipsTo}
        onChangeAddress={jest.fn()}
      />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('default value', () => {
    const tree = rendererCreate(
      <CountrySelector
        address={{
          ...newAddress,
          country: { value: 'BRA' },
        }}
        shipsTo={shipsTo}
        onChangeAddress={jest.fn()}
      />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should sort the select', () => {
    const tree = rendererCreate(
      <CountrySelector
        address={{
          ...newAddress,
          country: { value: 'BRA' },
        }}
        shipsTo={shipsTo}
        onChangeAddress={jest.fn()}
      />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('react to change value', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <CountrySelector
        address={{
          ...newAddress,
          country: { value: 'BRA' },
        }}
        shipsTo={shipsTo}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'USA' } }

    wrapper.find('select').simulate('change', event)

    expect(handleChange).toHaveBeenCalled()
  })

  it('should clean all input fields when country changes', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <CountrySelector
        address={{
          ...newAddress,
          country: { value: 'BRA' },
          postalCode: { value: '123' },
          state: { value: 'Foo' },
          city: { value: 'Bar' },
          neighborhood: { value: 'Baz' },
        }}
        shipsTo={shipsTo}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'USA' } }

    wrapper.find('select').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      country: { value: 'USA' },
      city: { value: null },
      complement: { value: null },
      geoCoordinates: { value: [] },
      neighborhood: { value: null },
      number: { value: null },
      postalCode: { value: null },
      reference: { value: null },
      state: { value: null },
      street: { value: null },
    })
  })
})
