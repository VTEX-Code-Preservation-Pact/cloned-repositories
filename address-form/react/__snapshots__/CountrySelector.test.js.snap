// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector default value 1`] = `
<p
  className="input ship-country text"
>
  <label
    htmlFor="ship-country"
  >
    Country
  </label>
  <select
    className=""
    disabled={false}
    id="ship-country"
    name="country"
    onBlur={[Function]}
    onChange={[Function]}
    value="BRA"
  >
    <option
      value="BRA"
    >
      Brazil
    </option>
    <option
      value="USA"
    >
      United States of America
    </option>
  </select>
</p>
`;

exports[`CountrySelector should sort the select 1`] = `
<p
  className="input ship-country text"
>
  <label
    htmlFor="ship-country"
  >
    Country
  </label>
  <select
    className=""
    disabled={false}
    id="ship-country"
    name="country"
    onBlur={[Function]}
    onChange={[Function]}
    value="BRA"
  >
    <option
      value="BRA"
    >
      Brazil
    </option>
    <option
      value="USA"
    >
      United States of America
    </option>
  </select>
</p>
`;

exports[`CountrySelector show options 1`] = `
<p
  className="input ship-country text"
>
  <label
    htmlFor="ship-country"
  >
    Country
  </label>
  <select
    className=""
    disabled={false}
    id="ship-country"
    name="country"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  >
    <option
      value="BRA"
    >
      Brazil
    </option>
    <option
      value="USA"
    >
      United States of America
    </option>
  </select>
</p>
`;
